### DISCUSSION TITLE
Modeling a Scalable Subcategory Table

### DISCUSSION THREAD

**Discussion by masnwilliams** [prisma/prisma#11973] [Q&A] [answered]
Hello! I am trying to model a Category table that can relate to a category, a parent category, in the same table.

[This Stack Overflow link provides a good example](https://stackoverflow.com/questions/27010527/create-sql-subcategory-table)

My goal is to build a scalable category system that could theoretically go to an infinite depth. Here is something similar to what I am trying to achieve:
![image](https://user-images.githubusercontent.com/43387599/155074506-f788a2ec-62d8-4926-ab88-030fca93399c.png)

This was my attempt at what I thought as the logical way to model it:
```
model ProductCategory {
  id               Int              @id @default(autoincrement())
  createdAt        DateTime         @default(now())
  updatedAt        DateTime?        @updatedAt
  products         Product[]        @relation(references: [id])
  name             String
  description      String?          @db.Text
  tier             Int              @default(1)
  parentCategoryId Int?
  parentCategory   ProductCategory? @relation(references: [id])

  @@map("productCategories")
}
```
The model above results in this error on the `parentCategory` line:
```
Error validating field `parentCategory` in model `ProductCategory`: The relation field `parentCategory` on Model `ProductCategory` is missing an opposite relation field on the model `ProductCategory`. Either run `prisma format` or add it manually.
```

I could simple do this:
```
model ProductCategory {
  id               Int              @id @default(autoincrement())
  createdAt        DateTime         @default(now())
  updatedAt        DateTime?        @updatedAt
  products         Product[]        @relation(references: [id])
  name             String
  description      String?          @db.Text
  tier             Int              @default(1)
  parentCategoryId Int?

  @@map("productCategories")
}
```
This removes the error but I lose the object mapping of the parent category. 

I feel like there has to be a solution to this. Any help is extremely appreciated!

**Comment by nurul3101** [comment, marked as answer]
Hey @johnmasonwilliams 👋🏻 ,

Could you try this `ProductCategory` model?

```prisma
model ProductCategory {
  id               Int               @id @default(autoincrement())
  createdAt        DateTime          @default(now())
  updatedAt        DateTime?         @updatedAt
  products         Product[]         @relation(references: [id])
  name             String
  description      String?           @db.Text
  tier             Int               @default(1)
  parentCategoryId Int
  parentCategory   ProductCategory?  @relation("ProductCategoryToProductCategory", fields: [parentCategoryId], references: [id])
  productCategory  ProductCategory[] @relation("ProductCategoryToProductCategory")

  @@map("productCategories")
}
```

Here is a reference for [self relations in prisma schema](https://www.prisma.io/docs/concepts/components/prisma-schema/relations/self-relations) which might be useful for your use case.

Please let me know if this solves the issue for you.

**Reply by masnwilliams** [reply]
Thank you so much for the quick reply!

I plugged your code in and was given this error:
```
Error validating: A self-relation must have `onDelete` and `onUpdate` referential actions set to `NoAction` in one of the @relation attributes. Read more at https://pris.ly/d/cyclic-referential-actions
```
I added the `onDelete: NoAction, onUpdate: NoAction` and it works!

Final working self-relation table:
```
model ProductCategory {
  id               Int               @id @default(autoincrement())
  createdAt        DateTime          @default(now())
  updatedAt        DateTime?         @updatedAt
  products         Product[]         @relation(references: [id])
  name             String
  description      String?           @db.Text
  tier             Int               @default(1)
  parentCategoryId Int
  parentCategory   ProductCategory?  @relation("ProductCategoryToProductCategory", fields: [parentCategoryId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  productCategory  ProductCategory[] @relation("ProductCategoryToProductCategory")

  @@unique([name, parentCategoryId])
  @@index([tier])
  @@map("productCategories")
}
```

Thank you for the link to the self relations as well.

Thank you!

**Reply by nurul3101** [reply]
Awesome! Thanks for mentioning about needing referential actions for self relations!

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
